### PR TITLE
For #12079 - Add 'additionalValidation' parameter to context menu options builders

### PR DESCRIPTION
**Pull request by Mugurell**
This will allow clients know when the context menu is opened and what options
it contains while also allowing clients to lazily control what options are
shown without the need for complex if-else conditions when the menu is set up.

Tests copy the already existing tests where the candidate is checked to be
shown by default and check that `additionalValidation` is called and can change
that.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Review by ionutbedregeanu — APPROVED** [review]

**Comment by Mugurell** [comment]
Thank you for the review @ionutbedregeanu .
Also checked with @Amejia481 who recommended to wait until the beta cut to land this.
Will update to add documentation for each method to make the new parameter clearer in it's intent until then.

**Comment by Mugurell** [comment]
Rebased to fix changelog conflicts.
With a new release started I guess this is also ready to land.

**Pull request merged**
